### ISSUE TITLE
IEEE754 non-conformance: is_sign_negative does not apply on NANs

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#42425]
IEEE754 says

> ― boolean isSignMinus(source)
isSignMinus(x) is true if and only if x has negative sign. isSignMinus applies to zeros and NaNs
as well.

However our `is_sign_negative`/`is_sign_positive` do not properly inspect the sign for `NaN`s.

**Comment by Thiez** [comment]
This program:

```rust
fn main() {
    let nums = [::std::f64::NEG_INFINITY, -1.0, -0.0, ::std::f64::NAN, 0.0, 1.0, ::std::f64::INFINITY];
    for &num in &nums[..] {
        println!("Number: {}\tSign negative: {}\tSign positive: {}", num, num.is_sign_negative(), num.is_sign_positive());
    }
}
```

prints the following output when compiled with current master and nightly:

```
Number: -inf	Sign negative: true	Sign positive: false
Number: -1	Sign negative: true	Sign positive: false
Number: 0	Sign negative: true	Sign positive: false
Number: NaN	Sign negative: false	Sign positive: false
Number: 0	Sign negative: false	Sign positive: true
Number: 1	Sign negative: false	Sign positive: true
Number: inf	Sign negative: false	Sign positive: true
```

What would be the expected result according to the standard? I would expect something that isn't a number to be neither negative nor positive, so I think the current results are correct?

**Comment by nagisa** [comment]
My reading of the spec suggests the answer should be equivalent to `is_sign_negative = signbit(x) != 0`, where `signbit` is non-zero if the binary representation has sign bit set.

E.g. Section 6.2.1:

> When encoded, all NaNs have a sign bit and a pattern of bits necessary to identify the encoding as a NaN and which determines its kind (sNaN vs. qNaN).

alludes to `NaN`s having a sign bit. The description of `isSignMinus` seems to quite clearly signal towards the fact that for `NaN`s the implementation ought to be inspecting this sign bit.

You can set the sign of `NaN` in rust with `-::std::f64::NAN` as well.

**Comment by rkruppe** [comment]
Isn't this fixed since #42431 was merged?
